### PR TITLE
Running, sitting and AFK anim. states

### PR DESCRIPTION
**Pull request by VoxelDoesCode**
Added in two more animation states, for running at a fast enough speed and for when you're AFK. Both of these states are bi-directional, and actually mirror very well inside the code.

The running animation I feel like gives a better explanation for how fast you're going, and stopping now would result in skidding. 
Old run cycle, which is just really fast walk:
https://user-images.githubusercontent.com/95713843/188262126-a2f63f92-88b5-408e-a1dd-af9288541c54.mp4
New run cycle:
https://user-images.githubusercontent.com/95713843/188262128-d7734eb7-fd43-4a6d-8605-c76b78ac759b.mp4

An AFK tee will also sit down and put its weapon behind its back; if not firing it. Very fitting with the many chairs and tables in maps.
unknown

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
I like it. :)

File: src/game/client/components/players.cpp [hunk 1]
```diff
@@ -454,8 +466,27 @@ void CPlayers::RenderPlayer(
...
+			case WEAPON_HAMMER: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_HammerHitFireDelay / 1000.0f; break;
+			case WEAPON_GUN: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_GunFireDelay / 1000.0f; break;
+			case WEAPON_SHOTGUN: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_ShotgunFireDelay / 1000.0f; break;
+			case WEAPON_GRENADE: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_GrenadeFireDelay / 1000.0f; break;
+			case WEAPON_LASER: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_LaserFireDelay / 1000.0f; break;
+			case WEAPON_NINJA: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_NinjaFireDelay / 1000.0f; break;
+			default: Delay = 1.0f;
+			}
+
+			bool IsActive = !m_pClient->m_aClients[ClientID].m_Afk || LastAttackTime < Delay;
```
Comment (Chairn) [inline src/game/client/components/players.cpp:483]: Is there a risk that the comparison may alternate ?

File: src/game/client/components/players.cpp [hunk 2]
```diff
@@ -454,8 +466,27 @@ void CPlayers::RenderPlayer(
...
+
+			switch(Player.m_Weapon)
+			{
+			case WEAPON_HAMMER: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_HammerHitFireDelay / 1000.0f; break;
+			case WEAPON_GUN: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_GunFireDelay / 1000.0f; break;
+			case WEAPON_SHOTGUN: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_ShotgunFireDelay / 1000.0f; break;
+			case WEAPON_GRENADE: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_GrenadeFireDelay / 1000.0f; break;
+			case WEAPON_LASER: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_LaserFireDelay / 1000.0f; break;
+			case WEAPON_NINJA: Delay = (float)m_pClient->m_aTuning[g_Config.m_ClDummy].m_NinjaFireDelay / 1000.0f; break;
+			default: Delay = 1.0f;
```
Comment (Robyt3) [inline src/game/client/components/players.cpp:480]: You could extract this into a method `CTuningParams::GetWeaponFireDelay(int Weapon)`. And instead of returning `1.0f` in the default case do a `dbg_assert(false, "invalid weapon")`. I think using a switch case is fine for this. Using an array or the weapon indices would be less readable and also more likely to break if the order of weapons or tuning parameters ever changes.

File: src/game/client/components/players.cpp (hunk 1 above)
Comment (VoxelDoesCode) [reply src/game/client/components/players.cpp:483]: Well, It shouldn't. Unless you're holding the fire button and SOMEHOW the LastAttackTime goes over the given weapon Delay.

**Comment by VoxelDoesCode** [comment]
There is a new tab in Appearance, called Animations, which will toggle on and off special animation states. By default they are deactivated, but can easily be turned on.
image

**Comment by heinrich5991** [comment]
I would rather not add options for this.

**Comment by archimede67** [comment]
> I would rather not add options for this.

Not everyone may want this behaviour, I think it does not hurt to have the choice

File: src/game/client/components/players.cpp [hunk 3]
```diff
@@ -410,22 +412,33 @@ void CPlayers::RenderPlayer(
 
 	bool Stationary = Player.m_VelX <= 1 && Player.m_VelX >= -1;
 	bool InAir = !Collision()->CheckPoint(Player.m_X, Player.m_Y + 16);
+	bool Running = Player.m_VelX <= -5000 || Player.m_VelX >= 5000;
```
Comment (Robyt3) [inline src/game/client/components/players.cpp:415]:
```suggestion
	bool Running = absolute(Player.m_VelX) >= 5000;
```

File: src/game/gamecore.h [hunk 4]
```diff
@@ -66,6 +66,7 @@ class CTuningParams
 	bool Get(const char *pName, float *pValue) const;
 	static const char *Name(int Index) { return ms_apNames[Index]; }
 	int PossibleTunings(const char *pStr, IConsole::FPossibleCallback pfnCallback = IConsole::EmptyPossibleCommandCallback, void *pUser = nullptr);
+	float GetWeaponFireDelay(int Weapon);
```
Comment (Robyt3) [inline src/game/gamecore.h:69]: The method could be declared `const`, like the `Get` method.

**Comment by VoxelDoesCode** [comment]
> I would rather not add options for this.
It might be extremely jarring for current players to just suddenly see tees act differently; especially the running. And, who knows, maybe someone might add more into animation!

**Comment by heinrich5991** [comment]
> It might be extremely jarring for current players to just suddenly see tees act differently; especially the running. And, who knows, maybe someone might add more into animation!

I'm against more config options, especially those exposed to the UI.

Perhaps we could try introducing some global "I want new UI" setting that contains all the "chat tees", etc. and this?

File: src/game/gamecore.cpp [hunk 5]
```diff
@@ -7,6 +7,7 @@
 #include "teamscore.h"
 
 #include <engine/shared/config.h>
+#include <game/client/gameclient.h>
```
Comment (Robyt3) [inline src/game/gamecore.cpp:10]:
```suggestion
```

**Comment by fokkonaut** [comment]
> > It might be extremely jarring for current players to just suddenly see tees act differently; especially the running. And, who knows, maybe someone might add more into animation!
> 
> I'm against more config options, especially those exposed to the UI.
> 
> Perhaps we could try introducing some global "I want new UI" setting that contains all the "chat tees", etc. and this?

Please no. Not everyone wants everything, but still like a few selected features. Instead we maybe should use some bitwise config settings. But not one for everything or nothing.

**Comment by heinrich5991** [comment]
> Please no. Not everyone wants everything, but still like a few selected features. Instead we maybe should use some bitwise config settings. But not one for everything or nothing.

The problem with that is maintenance overhead. Developing new features gets harder with every config flag that affects your new feature. Every config option is expected by players to work with every other config option. I'd prefer leaving that amount of customizability to a custom client or it not existing at all if there's no such custom client. Do you want to discuss this further? Let's create a separate discussion issue.

**Comment by VoxelDoesCode** [comment]
So the only thing I need to do now is remove the toggle on/off settings and then it's good to go?

**Comment by heinrich5991** [comment]
I'd give others a day or two to comment on the animation itself, but yes, otherwise it's good to go. :)

**Comment by C0D3D3V** [comment]
I have not tested it yet. But it looks good. 

What about skidding? If you run with a lot of speed and then start walking in the oposide direction? Before there was no feet animation so it looks like you stopping with your feets.  (with `else if(!WantOtherDir)`)
Do you keep that behaviour?

**Comment by VoxelDoesCode** [comment]
> What about skidding? If you run with a lot of speed and then start walking in the opposite direction? Before there was no feet animation so it looks like you stopping with your feet. (with else if(!WantOtherDir))
Do you keep that behavior?

Yeah, that's kept.

**Comment by C0D3D3V** [comment]
https://www.dailymotion.com/video/x8dj5fz  pw: 1234

https://discord.com/channels/252358080522747904/293493549758939136/1017056431964434432

From Discord by Ravie:
> The weapon holding looks weird, especially hammer
> I think [position of the feet is] a little too far to the side
> also when facing to the left, the front foot is upside down xD, dunno a lot of things about this look weird to me

From fokkonaut:
> maybe we could use the same [animation] for freeze btw, with weapons on back and sitting
> right now the tee doesnt have any weapo

Any comments @VoxelDoesCode? I personally think it looks fine.

**Comment by C0D3D3V** [comment]
What Ravie means with "also when facing to the left, the front foot is upside down xD, dunno a lot of things about this look weird to me":

grafik

grafik

**Comment by VoxelDoesCode** [comment]
> Any comments @VoxelDoesCode? I personally think it looks fine.

Well, let's test a few things out before I call it final.

### TODO

- [ ] Try out new sitting/weapon positions
- [ ] Implement foot flipping
- [ ] Try using sit animation for freeze

**Comment by VoxelDoesCode** [comment]
Update: Feet now point up less and gun simply aims down. Feet have the ability to flip so that they don't go topsy-turvy when looking to the left. Now running and walking look natural as well when it comes to unique skins! Not sure if I missed anything, but I got most of what I needed to do. Thanks to c0d3d3v and Ravie for boosting this PR!

https://user-images.githubusercontent.com/95713843/189015435-f2ef0b7b-07f4-4e6e-a1c8-ee563a4b185e.mp4

There is a temporary toggle named cl_force_afk, which will be removed when everything is validated. This is for ease of testing.

> maybe we could use the same [animation] for freeze btw, with weapons on back and sitting

As for this, I tried it and it really didn't feel too fitting, especially with the gun angle.

File: src/game/server/entities/character.cpp [hunk 6]
```diff
@@ -2257,6 +2262,7 @@ void CCharacter::DDRaceInit()
 	m_TeamBeforeSuper = 0;
 	m_Core.m_Id = GetPlayer()->GetCID();
 	m_TeleCheckpoint = 0;
+	m_IsSitting =
```
Comment (fokkonaut) [inline src/game/server/entities/character.cpp:2265]: m_IsSitting = false;

**Comment by fokkonaut** [comment]
I like this a lot. The manual sitting is fun. A few things I noticed:
- The hammer going up and down is very nice, I think we should do it for all weapons (I'm not sure if those 3 subtiles make a huge difference for the projectiles)
- If the server reports you are in sitting mode, but you get dragged so your tee gets out of sitting mode on the clientside, the weapon does not go up with it. Would require a vel check probably? 
- Jumping should maybe also reset sitting state, that could simply be added to the Direction != 0 check on the server

https://user-images.githubusercontent.com/35420825/189342881-11c4db67-37de-4623-9beb-173255fd48cc.mp4

**Comment by heinrich5991** [comment]
Adding a way to manually sit outside of AFK (and /pause) removes the signal that sitting would otherwise have.

I'd prefer to make sitting such a strong signal. You can still manually sit if you wanted to by going into /pause. Obviously, you couldn't react as fast as with the sit command. But that's basically what the signal would be about.

**Comment by C0D3D3V** [comment]
I also do not like the command. Starting with "dance" or " cute static" animations should be part of an other PR. If we would add such animations there should be a selection wheel for it like there is for emoticons and eyes.

**Comment by VoxelDoesCode** [comment]
Alright; So I have decided: I will not make sitting a manual thing now, and will revert to just /pause and AFK state.
As fun as it might be to be able to sit on will, it's very clear that I have a lot to learn in terms of game design, and my ideas might be too bold for the game's future.
It will take a bit to revert everything, but hopefully, once we talk about this idea more, we can bring it back (in one way or another).

**Comment by C0D3D3V** [comment]
I noticed that if there is a boot with afk player at the bootom. The feet goes back and forth... that looks strange.

**Comment by VoxelDoesCode** [comment]
I mean, I think the feet flickering was always like that? I'm not sure.

**Comment by C0D3D3V** [comment]
Mh, good point. I had the old animation in my build maybe with the new the wiggle is not that strong visible. I will make a comparison between new animation and no animation :D

**Comment by heinrich5991** [comment]
Status: Still active discussion, will wait before merging.

**Comment by def-** [comment]
What is left to discuss? I think this would be good to go in for a new DDNet 16.4 release

**Comment by heinrich5991** [comment]
I saw this:

@C0D3D3V wrote:
> I noticed that when there is a boat with an afk player on the ground, that player's feet wiggle wildly back and forth... it looks weird.

@VoxelDoesCode wrote:
> I mean, I think the feet flickering was always like that? I'm not sure.

@C0D3D3V wrote:
> Mh, good point. I had the old animation in my build maybe with the new the wiggle is not that strong visible. I will make a comparison between new animation and no animation :D

---

 I'll review the PR and assume the discussion above is finished.

File: src/game/client/components/players.cpp [hunk 7]
```diff
@@ -408,26 +410,45 @@ void CPlayers::RenderPlayer(
 
 	RenderInfo.m_GotAirJump = Player.m_Jumped & 2 ? 0 : 1;
 
-	bool Stationary = Player.m_VelX <= 1 && Player.m_VelX >= -1;
+	bool Stationary = absolute(Player.m_VelX) <= 1;
```
Comment (heinrich5991) [inline src/game/client/components/players.cpp:413]: `absolute` is weird about the minimum integer, it just returns the minimum integer for it, still negative (technically, it's undefined behavior but the compiler doesn't seem to recognize this). This leads to bad code generation here, so that the old code is actually faster. Could you undo the change? I don't find the new code much more readable.

File: src/game/client/components/players.cpp [hunk 8]
```diff
@@ -632,10 +665,10 @@ void CPlayers::RenderPlayer(
 				Client()->IntraGameTick(g_Config.m_ClDummy));
 
 		CTeeRenderInfo Shadow = RenderInfo;
-		RenderTools()->RenderTee(&State, &Shadow, Player.m_Emote, Direction, ShadowPosition, 0.5f); // render ghost
+		RenderTools()->RenderTee(&State, &Shadow, Player.m_Emote, Direction, ShadowPosition, 0.5f, FlipFeet); // render ghost
 	}
 
-	RenderTools()->RenderTee(&State, &RenderInfo, Player.m_Emote, Direction, Position, Alpha);
+	RenderTools()->RenderTee(&State, &RenderInfo, Player.m_Emote, Direction, Position, Alpha, FlipFeet);
```
Comment (heinrich5991) [inline src/game/client/components/players.cpp:671]: I think the `FlipFeet` belongs into the `RenderInfo`. Could you add it to the `m_TeeRenderFlags`? Adding a parameter for every new detail wouldn't scale.

File: src/game/client/components/menus_settings.cpp [hunk 9]
```diff
@@ -1544,7 +1544,7 @@ void CMenus::RenderSettingsGraphics(CUIRect MainView)
 	CUIRect Text;
 	MainView.HSplitTop(20.0f, 0, &MainView);
 	MainView.HSplitTop(20.0f, &Text, &MainView);
-	//text.VSplitLeft(15.0f, 0, &text);
+	// text.VSplitLeft(15.0f, 0, &text);
```
Comment (heinrich5991) [inline src/game/client/components/menus_settings.cpp:1547]: In general, avoid touching code that doesn't belong to your changes.

**Review by heinrich5991 — COMMENTED** [review]
Looks good, I want to have this. I added two minor nits.

I'd also this PR to be squashed. Should I do this? @VoxelDoesCode Then I can also address the two review comments I left.

File: src/game/client/components/players.cpp (hunk 7 above)
Comment (VoxelDoesCode) [reply src/game/client/components/players.cpp:413]: Should this be changed for the running speed as well?

Comment (heinrich5991) [reply src/game/client/components/players.cpp:413]: I guess.

File: src/game/client/components/players.cpp [hunk 10]
```diff
@@ -408,26 +410,46 @@ void CPlayers::RenderPlayer(
 
 	RenderInfo.m_GotAirJump = Player.m_Jumped & 2 ? 0 : 1;
 
+	RenderInfo.m_FeetFlipped = false;
+
 	bool Stationary = Player.m_VelX <= 1 && Player.m_VelX >= -1;
 	bool InAir = !Collision()->CheckPoint(Player.m_X, Player.m_Y + 16);
+	bool Running = Player.m_VelX >= 5000 || Player.m_VelX <= -5000;
```
Comment (edg-l) [inline src/game/client/components/players.cpp:417]: I would add a constant instead of the 5000, maybe TEE_RUNNING_THRESHOLD is a good name for it.

**Review by edg-l — COMMENTED** [review]
minor nitpick, otherwise LGTM

File: src/game/client/components/players.cpp (hunk 10 above)
Comment (def-) [reply src/game/client/components/players.cpp:417]: I disagree, it's only used in this one spot and being used to define the variable `Running` tells you what the value does.

**Comment by def-** [comment]
Squashed
Thanks @VoxelDoesCode for the cool new features :)
bors r+
